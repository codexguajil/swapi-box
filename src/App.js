import React, { Component } from 'react';
import { FilmScroller } from './FilmScroller';
import { Header } from './Header';
import CardContainer from './CardContainer';
import './App.css';

class App extends Component {
  constructor() {
    super();
    this.state={
      films: [],
      count: 7,
    }
  }

  componentDidMount() {
    const url = 'https://swapi.co/api/films/'
    fetch(url)
      .then(response => response.json())
      .then(data => this.setState({films: data.results}))
  }

  render() {
    console.log(this.state)
    const { films } = this.state
    return (
      <div className="App">
        <Header />
        <CardContainer />
        <FilmScroller films={films}/>
      </div>
    );
  }
}

export default App;
