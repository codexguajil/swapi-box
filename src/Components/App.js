import React, { Component } from 'react';
import { FilmScroller } from './FilmScroller';
import { Header } from './Header';
import { CardsContainer } from './CardsContainer';
import { ButtonsContainer } from './ButtonsContainer';
import { fetchFilms, fetchApi, iterateFetch } from '../Methods/apiCalls';
import { iterateHelper } from '../Methods/helpers';
import './App.css';
import { promises } from 'fs';

class App extends Component {
  constructor() {
    super();
    this.state={
      films: [],
      count: 0,
      data: [],
    }
    this.fetchPeople = this.fetchPeople.bind(this)
    this.fetchSpecies = this.fetchSpecies.bind(this)
  }

  componentDidMount() {
    const url = 'https://swapi.co/api/films/'
    this.fetchFilms(url)
  }

  async fetchFilms(url) {
    try {
      const films = await fetchFilms(url)
      this.setState({films, count: films.length})
    } catch(error) {
      console.log(error.message)
    }
  }

  async fetchPeople(value) {  
    console.log(value)
    const url = `https://swapi.co/api/${value}`
    try {
      const parsedResponse = await fetchApi(url)
      const objectValues = await iterateHelper(parsedResponse.results, value)
      console.log('objectVAlues:', objectValues)
      if(objectValues[0].species) {
        let homeWorlds = await this.fetchHomeWorlds(objectValues)
        let species = await this.fetchSpecies(objectValues)
        homeWorlds.forEach((person,i) => {
          person.species = species[i].name
          person.language = species[i].language
        })
        console.log(homeWorlds);
        this.setState({data: homeWorlds})
        } else if(objectValues[0].terrain) {
          let residents = await this.fetchResidents(objectValues)
          let newPlanets = this.resetResidents(objectValues, residents)
          console.log(newPlanets)
          this.setState({data: newPlanets})
        } else if(objectValues[0].model) {
          this.setState({data: objectValues})
        }
      } catch(error) {
        console.log(error.message)
      }
    }

  resetResidents = (planetValues, residents) => {
    let resetPlanetValues = planetValues.map((planet,i) => {
      return {...planet, residents: residents[i]}
    })
    return resetPlanetValues
  }

  async fetchResidents(planets) {
    try {
      let newPlanets = planets.map(planet => {
        let fetchedResidents = this.fetchUrls(planet.residents)
        return fetchedResidents
      })
      return await Promise.all(newPlanets)
    }catch(error) {
      console.log(error.message)
    }
  }

  async fetchUrls(urls) {
    if(urls.length > 0) {
      let fetchedUrls = await urls.map(url => {
        return fetchApi(url)
      })
      return await Promise.all(fetchedUrls)
    } else {
      urls.push({name:'unknown'})
      return urls;
    }
  }

  async fetchHomeWorlds(people) {
    console.log('people', people)
    let homes = people.map(person => {
       return fetchApi(person.home)
    })
    let allHomes = await Promise.all(homes)
    let homeValues = allHomes.map((home,i) => {
      return {
        name: people[i].name,
        Homeworld: home.name,
        Population: home.population
      }
    })

    return homeValues
  }

  async fetchSpecies(people) {
    console.log(people)
    let species = people.map(person => {
      return fetchApi(person.species)
    })
    let allSpecies = await Promise.all(species)
    let speciesName = allSpecies.map(species => {
      return {
        name: species.name,
        language: species.language
      }
    })
    return speciesName;
  }

  render() {
    const { films, data } = this.state
    return (
      <div className="App">
        <Header />
        <ButtonsContainer fetchPeople={this.fetchPeople}/>
        <CardsContainer data={data}/>
        <FilmScroller films={films}/>
      </div>
    );
  }
}

export default App;
