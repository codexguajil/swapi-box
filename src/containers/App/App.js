import React, { Component } from 'react';
import FilmScroller from '../FilmScroller/FilmScroller';
import { fetchAnything } from '../../utils/apiCalls';
import { connect } from 'react-redux';
import { addMovie } from '../../actions';
import { ButtonsContainer } from '../../Components/ButtonsContainer';
import { Header } from '../../Components/Header';
import CardsContainer from '../CardsContainer/CardsContainer';

export class App extends Component {

  componentDidMount = () => {
    this.fetchRandomFilm();
  }
  
  fetchRandomFilm = async () => {
    const url = 'https://swapi.co/api/films/';
    const randomNum = Math.floor(Math.random() * 6);
    try {
      const films = await fetchAnything(url);
      this.props.addMovie(films.results[randomNum]);
    } catch(error) {
      throw error.message;
    }
  }

  render() {
    return (
        <div className="App">
          <Header />
          <FilmScroller />
          <ButtonsContainer />
          <CardsContainer />
        </div>
    );
  }
}

export const mapStateToProps = (state) => ({
  movie: state.movie,
  people: state.people,
})

export const mapDispatchToProps = (dispatch) => ({
  addMovie: (movie) => dispatch(addMovie(movie))
})

export default connect(mapStateToProps, mapDispatchToProps)(App);