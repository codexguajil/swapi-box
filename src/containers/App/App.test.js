import React from 'react';
import {shallow} from 'enzyme';
import {App} from './App'; 

describe('App', () => {
  let wrapper;
  let mockFn;

  beforeEach(() => {
    mockFn = jest.fn()
    wrapper = shallow(
      <App />
    )
  })

  it('should match snapshot', () => {

    expect(wrapper).toMatchSnapshot()
  })

  it('should call fetchRandomFilm upon mounting', () => {
    const wrapper = shallow(<App />)
    const instance = wrapper.instance();
    jest.spyOn(instance, 'fetchRandomFilm');
    instance.componentDidMount();

    expect(instance.fetchRandomFilm).toHaveBeenCalled();
    mockFn.mockClear()
  })
})