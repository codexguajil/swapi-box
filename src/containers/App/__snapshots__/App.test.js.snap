// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot 1`] = `
<div
  className="App"
>
  <Header />
  <Connect(FilmScroller) />
  <ButtonsContainer />
  <Connect(CardsContainer) />
</div>
`;
